internal/wrapper.go

```diff
@@ -30,67 +29,73 @@ import (
 	"k8s.io/klog/v2"
 
 	"k8s.io/kube-state-metrics/v2/pkg/app"
-	"k8s.io/kube-state-metrics/v2/pkg/customresource"
-	"k8s.io/kube-state-metrics/v2/pkg/customresourcestate"
 	"k8s.io/kube-state-metrics/v2/pkg/options"
 )
 
 // RunKubeStateMetricsWrapper is a wrapper around KSM, delegated to the root command.
 func RunKubeStateMetricsWrapper(opts *options.Options) {
-	var factories []customresource.RegistryFactory
-	if config, set := resolveCustomResourceConfig(opts); set {
-		crf, err := customresourcestate.FromConfig(config)
-		if err != nil {
-			klog.ErrorS(err, "Parsing from Custom Resource State Metrics file failed")
-			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
-		}
-		factories = append(factories, crf...)
-	}
 
 	KSMRunOrDie := func(ctx context.Context) {
-		if err := app.RunKubeStateMetricsWrapper(ctx, opts, factories...); err != nil {
+		if err := app.RunKubeStateMetricsWrapper(ctx, opts); err != nil {
 			klog.ErrorS(err, "Failed to run kube-state-metrics")
 			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
 		}
 	}
+
 	ctx, cancel := context.WithCancel(context.Background())
 	if file := options.GetConfigFile(*opts); file != "" {
-		viper.SetConfigType("yaml")
-		viper.SetConfigFile(file)
-		if err := viper.ReadInConfig(); err != nil {
+		cfgViper := viper.New()
+		cfgViper.SetConfigType("yaml")
+		cfgViper.SetConfigFile(file)
+		if err := cfgViper.ReadInConfig(); err != nil {
 			if errors.Is(err, viper.ConfigFileNotFoundError{}) {
 				klog.ErrorS(err, "Options configuration file not found", "file", file)
 			} else {
 				klog.ErrorS(err, "Error reading options configuration file", "file", file)
 			}
 			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
 		}
-		viper.OnConfigChange(func(e fsnotify.Event) {
+		cfgViper.OnConfigChange(func(e fsnotify.Event) {
 			klog.Infof("Changes detected: %s\n", e.Name)
 			cancel()
 			// Wait for the ports to be released.
 			<-time.After(3 * time.Second)
 			ctx, cancel = context.WithCancel(context.Background())
 			go KSMRunOrDie(ctx)
 		})
-		viper.WatchConfig()
-	}
-	klog.Infoln("Starting kube-state-metrics")
-	KSMRunOrDie(ctx)
-	select {}
-}
+		cfgViper.WatchConfig()
 
-func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.ConfigDecoder, bool) {
-	if s := opts.CustomResourceConfig; s != "" {
-		return yaml.NewDecoder(strings.NewReader(s)), true
-	}
-	if file := opts.CustomResourceConfigFile; file != "" {
-		f, err := os.Open(filepath.Clean(file))
+		// Merge configFile values with opts so we get the CustomResourceConfigFile from config as well
+		configFile, err := os.ReadFile(filepath.Clean(file))
 		if err != nil {
-			klog.ErrorS(err, "Custom Resource State Metrics file could not be opened")
+			klog.ErrorS(err, "failed to read options configuration file", "file", file)
+		}
+
+		yaml.Unmarshal(configFile, opts)
```

Comment:
I believe this would require further logic (as it could also be set via command line), so merging seemed easier (to use the same logic that we have in https://github.com/kubernetes/kube-state-metrics/blob/b7a7070e87ba4299a90587c8c2d5c7647db1ba42/pkg/app/server.go#L129

Reply:
sounds good :+1: